xmtp_mls/src/identity_updates.rs

```diff
@@ -115,28 +116,37 @@ where
         inbox_id: InboxId,
         to_sequence_id: Option<i64>,
     ) -> Result<AssociationState, ClientError> {
-        // TODO: Check against a local cache before talking to the network
-
+        let inbox_id = inbox_id.as_ref();
         let updates = conn.get_identity_updates(inbox_id, None, to_sequence_id)?;
-        let last_update = updates.last();
-        if last_update.is_none() {
+        let last_sequence_id = updates
+            .last()
+            .ok_or::<ClientError>(AssociationError::MissingIdentityUpdate.into())?
+            .sequence_id;
+        if to_sequence_id.is_some() && to_sequence_id != Some(last_sequence_id) {
+            // Is this right?
```

Comment:
I tidied this a little bit, but the logic is still the same. I'm unsure of the assumptions behind the original logic - shouldn't we be tolerant if, for example, the storage is *ahead* of the requested sequence ID?

Also, we are putting some invisible requirements on the caller of this method to make sure that the requested `to_sequence_id` is in the DB, and to call `load_identity_updates` if not. Should we just refactor to call `load_identity_updates` natively within this method if needed?

Reply:
> shouldn't we be tolerant if, for example, the storage is ahead of the requested sequence ID?

If the storage is ahead, those entries shouldn't be returned because we are providing a `to_sequence_id` in the call to `conn.get_identity_updates`.

> Also, we are putting some invisible requirements on the caller of this method to make sure that the requested to_sequence_id is in the DB, and to call load_identity_updates if not

I'm sure there are cases for a combined `get_association_state` + `load_identity_updates`. That would tell you what an inboxes state is right now.

I was primarily thinking about commit processing when I wrote this.

For commit processing, we actually want to load all the updates for all changed inboxes in a batch operation to save trips to the network, and then we want to get diffs for each inbox between the old and new state, which involves getting the association state at one or two different sequence IDs (depending on whether someone is a new addition to the group or an update).

Reply:
> If the storage is ahead, those entries shouldn't be returned because we are providing a to_sequence_id in the call to conn.get_identity_updates

D'oh! That's obvious in hindsight.

> For commit processing, we actually want to load all the updates for all changed inboxes in a batch operation to save trips to the network, and then we want to get diffs for each inbox between the old and new state, which involves getting the association state at one or two different sequence IDs

Ahh I see. I'm wondering if it makes sense to pass the identity updates into the function, instead of separately fetching it in the function and validating against `to_sequence_id`, seeing as we're hitting the DB for `filter_inbox_ids_needing_updates()` anyway. But this is pretty minor at this point, thanks for explaining!